subprojects/resources-gcs/src/main/java/org/gradle/internal/resource/transport/gcp/gcs/GcsClient.java

```diff
@@ -0,0 +1,180 @@
+            // Setting the length improves upload performance
+            contentStream.setLength(contentLength);
+
+            // TODO - set ACL here if necessary
+            String bucket = destination.getHost();
+            String path = cleanResourcePath(destination);
+            StorageObject objectMetadata = new StorageObject().setName(path);
+
+            Storage.Objects.Insert putRequest = storage.objects().insert(bucket, objectMetadata, contentStream);
+
+            LOGGER.debug("Attempting to put resource:[{}] into gcs bucket [{}]", putRequest.getName(), putRequest.getBucket());
+            putRequest.execute();
+        } catch (IOException e) {
+            throw ResourceExceptions.putFailed(destination, e);
+        }
+    }
+
+    @Nullable
+    public StorageObject getResource(URI uri) throws ResourceException {
+        LOGGER.debug("Attempting to get gcs resource: [{}]", uri.toString());
+
+        String path = cleanResourcePath(uri);
+        try {
+            Storage.Objects.Get getRequest = storage.objects().get(uri.getHost(), path);
+            return getRequest.execute();
+        } catch (GoogleJsonResponseException e) {
+            if (e.getStatusCode() == 404) {
+                return null;
+            }
+            throw ResourceExceptions.getFailed(uri, e);
+        } catch (IOException e) {
+            throw ResourceExceptions.getFailed(uri, e);
+        }
+    }
+
+    @Nullable
+    public List<String> list(URI uri) throws ResourceException {
+        List<StorageObject> results = new ArrayList<StorageObject>();
+
+        try {
+            Storage.Objects.List listRequest = storage.objects().list(uri.getHost());
+            Objects objects;
+
+            // Iterate through each page of results, and add them to our results list.
+            do {
+                objects = listRequest.execute();
+                // Add the items in this page of results to the list we'll return.
+                results.addAll(objects.getItems());
+
+                // Get the next page, in the next iteration of this loop.
+                listRequest.setPageToken(objects.getNextPageToken());
+            } while (null != objects.getNextPageToken());
+        } catch (IOException e) {
+            throw ResourceExceptions.getFailed(uri, e);
+        }
+
+        List<String> resultStrings = new ArrayList<String>();
+        for(StorageObject result : results) {
+            resultStrings.add(result.getName());
+        }
+
+        return resultStrings;
+    }
+
+    @VisibleForTesting
+    InputStream getResourceStream(StorageObject obj) throws IOException {
+        Storage.Objects.Get getObject = storage.objects().get(obj.getBucket(), obj.getName());
+        getObject.getMediaHttpDownloader().setDirectDownloadEnabled(false);
+        return getObject.executeMediaAsInputStream();
+    }
+
+    private static String cleanResourcePath(URI uri) {
```

Comment:
Since this methods does not use any state from the enclosing class what would be the benefit to keep it instanced?